### ISSUE TITLE
Wix bundle security issues - Special permissions

### ISSUE THREAD

**Issue by benp87** [wixtoolset/issues#6149]
* Which version of WiX are you building with?

3.11.2

* Which version of Visual Studio are you building with (if any)?

Visual Studio 2019 16.5.0

* Which version of the WiX Toolset Visual Studio Extension are you building with (if any)?

Wix Toolset extension 2019 1.0.0.4

* Which version of .NET are you building with?

4.6.2

* If the problem occurs when installing your packages built with WiX, what is the version of Windows the package is running on?

Windows 10, version 1903

* Describe the problem and the steps to reproduce it.

When running an installer bundle, the content of the bundle is being extracted to a temporary folder in TEMP directory, if running normally without admin permissions, the folder is in a user's temp directory, otherwise the folder is in "C:\Windows\Temp".

I can see a security issue here:
In case of a user's temp directory, a hacker can just changed the DLLs of that folder and make the user run a malicious DLLs when he runs the bundle. 
But when the folder is "C:\Windows\Temp", even though only admin suppose to have permissions there, a normal user still have "special permissions" which allows him to change the files there:
<img width="945" alt="SpecialPermissions" src="https://user-images.githubusercontent.com/13748110/77305465-b442ef80-6cfe-11ea-9477-95aa294b991b.PNG">
This also allow hacker to replace DLLs..

* Describe the behavior you expected and how it differed from the actual behavior.

When the bundle content is extracted to "C:\Windows\Temp":
1. Create a new folder in "C:\Windows\Temp".
2. Change folder's ACL (Access Control List) to admin.
3. "User" should not have "special access" - FILE_WRITE_DATA.

**Comment by robmen** [comment]
This is mitigated because Burn uses a random name for the folder in TEMP that is mathematically impossible to guess by an attacker.

**Comment by benp87** [comment]
> 
> 
> This is mitigated because Burn uses a random name for the folder in TEMP that is mathematically impossible to guess by an attacker.

I can write a program that monitors the TEMP folder (Let's say using FileSystemWatcher .NET Class), immediately after a folder is created, I can replace the bootstrapper application DLL with my own DLL, before the burn loads it.

**Comment by robmen** [comment]
That will only work if the program monitoring the folder is elevated. Normal users can't scan the contents of the system TEMP folder.

**Comment by hathawcs** [comment]
@robmen 

If the bundle is not run as admin, the user's TEMP folder is used and not the system TEMP folder. A utility is able to monitor the user's TEMP folder for changes and drop its own DLL into the .be folder immediately when it is created. When the burn engine elevates, it will then be affected by the injected DLLs.

The solution would be to ensure the .be folder is created with, or is modified with, more restrictive permissions. Then, wipe the contents of the .be folder before elevating. This will protect installers from rogue software monitoring for the creation of the .be folder or those that inject their own .be folder (after monitoring for the .ba folder).

**Comment by robmen** [comment]
>  When the burn engine elevates, it will then be affected by the injected DLLs.

This is incorrect. The elevated engine protects against this. If you have a DLL that can be injected into the elevated engine without elevating, please do open a security issue.

> The solution would be to ensure the .be folder is created with, or is modified with, more restrictive permissions. 

This is illogical. If such a folder is created with permissions preventing the user from writing to it, how would the elevated engine be placed there?

**Comment by hathawcs** [comment]
I have a report of someone placing a modified COMCTL32.dll in a specific path within the .be folder before the elevated engine runs. The elevated engine then runs the modified COMCTL32.dll instead of the standard version.

I have confirmed the vulnerability.

> > The solution would be to ensure the .be folder is created with, or is modified with, more restrictive permissions.
> 
> This is illogical. If such a folder is created with permissions preventing the user from writing to it, how would the elevated engine be placed there?

Yes, this is illogical, but as a consumer of WiX, I don't have many options in improving how the elevated engine is placed and protected. Because the report was regarding a subfolder of .be, I ensured the user could not create subfolders in the .be folder to block this specific attack vector.

For another data point, here is a different installer framework reporting a similar finding and their attempts to block the vulnerability: https://www.advancedinstaller.com/exe-local-directory-vulnerability-solution.html
